### PR TITLE
Fix AlertBox and add facemask support in widget editing

### PR DESCRIPTION
**Pull request by gettinToasty**
Daniel pointed me in the right direction that the facemask settings, which were not previously editable in slobs but were included in the api requests, were getting corrupted. 7750d10 is the commit that will prevent this from happening in the future and will also unbork existing user settings

### CODE REVIEW

File: app/services/widgets/settings/alert-box/index.ts [hunk 1]
```diff
@@ -211,6 +212,10 @@ export class AlertBoxService extends WidgetSettingsService<IAlertBoxData> {
       setting.image_href === '/images/gallery/default.gif'
         ? 'http://uploads.twitchalerts.com/image-defaults/1n9bK4w.gif'
         : setting.image_href;
+    const constrainedDuration =
+      Math.floor(setting.alert_duration / 1000) <= 300
+        ? Math.floor(setting.alert_duration / 1000)
+        : 300;
```
Comment (holiber) [inline app/services/widgets/settings/alert-box/index.ts]: const constrainedDuration = Math.floor(Math.max(300, setting.alert_duration / 1000))

File: app/components/widgets/AlertBox.vue [hunk 2]
```diff
@@ -13,7 +13,7 @@
       <span class="button button--default add-alert-button" @click="toggleAddAlertMenu()">{{ $t('Add Alert') }}</span>
       <div v-if="addAlertMenuOpen" class="add-alert-dropdown">
         <button
-          v-for="type in alertTypes"
+          v-for="type in alertTypes.filter(t => t !== 'facemasks')"
```
Comment (blackxored) [inline app/components/widgets/AlertBox.vue]: Would prefer this kind of logic (any logic, really) on a computed prop instead.

File: app/components/widgets/AlertBox.vue [hunk 3]
```diff
@@ -31,7 +31,10 @@
         <span class="left-accordion__title">{{ alertName(alert) }}</span>
       </div>
       <div class="left-accordion__input" v-if="wData.settings[alert]">
-        <validated-form  @input="save()"><toggle-input v-model="wData.settings[alert].enabled" /></validated-form>
+        <validated-form v-if="alert !== 'facemasks'" @input="save()"><toggle-input v-model="wData.settings[alert].enabled" /></validated-form>
```
Comment (blackxored) [inline app/components/widgets/AlertBox.vue]: Nitpick: remove brackets.

File: app/components/widgets/AlertBox.vue [hunk 4]
```diff
@@ -31,7 +31,10 @@
...
+        <validated-form v-if="alert === 'facemasks'" @input="handleFacemaskInput()">
```
Comment (blackxored) [inline app/components/widgets/AlertBox.vue]: See comment about extracting above. This also impacts the previous line.

File: app/components/widgets/AlertBox.vue [hunk 5]
```diff
@@ -89,7 +92,7 @@
       <v-form-group v-model="selectedVariation.settings.text.color2" :metadata="metadata.secondaryColor" />
     </div>
   </validated-form>
-  <validated-form slot="media-properties" key="media-properties" @input="save()" v-if="selectedVariation">
+  <validated-form slot="media-properties" key="media-properties" @input="handleFacemaskInput()" v-if="selectedVariation">
```
Comment (blackxored) [inline app/components/widgets/AlertBox.vue]: Nitpick: remove brackets.

File: app/components/widgets/AlertBox.vue.ts [hunk 6]
```diff
@@ -95,6 +98,8 @@ export default class AlertBox extends WidgetSettings<IAlertBoxData, AlertBoxServ
   editingName: string = null;
   languages: any[] = [];
 
+  facemaskEnabled = this.facemasksService.getEnabledStatus();
```
Comment (blackxored) [inline app/components/widgets/AlertBox.vue.ts]: If this doesn't change after init then it's fine as property initializer. Can you please confirm that's the case? Thanks.

File: app/components/widgets/AlertBox.vue.ts [hunk 7]
```diff
@@ -203,4 +208,18 @@ export default class AlertBox extends WidgetSettings<IAlertBoxData, AlertBoxServ
...
+  handleFacemaskInput() {
+    if (this.selectedAlert === 'facemasks') {
+      const { duration } = this.selectedVariation.settings;
+      this.facemasksService
+        .updateFacemaskSettings({
+          duration,
+          enabled: this.facemaskEnabled,
+          device: this.facemasksService.getEnabledDevice(),
+        })
+        .catch(() => this.onFailHandler($t('Something went wrong updating Facemask settings')));
```
Comment (blackxored) [inline app/components/widgets/AlertBox.vue.ts]: Do we still see value in reporting to Sentry? 
`Sentry.captureException` prior to invoking `onFailHandler`? Please disregard if the handler already does this.

File: app/services/widgets/settings/alert-box/index.ts (hunk 1 above)
Comment (gettinToasty) [reply app/services/widgets/settings/alert-box/index.ts]: i dont want to hold up preview so i'll make a note of that and make this change in a future PR

File: app/services/facemasks.ts [hunk 8]
```diff
@@ -399,9 +420,7 @@ export class FacemasksService extends PersistentStatefulService<IFacemasksServic
     return new Promise((resolve, reject) => {
       const asyncReads = missing.map(uuid => this.readFile(uuid));
       Promise.all(asyncReads)
-        .then(results => {
-          resolve();
-        })
+        .then(results => resolve())
```
Comment (blackxored) [inline app/services/facemasks.ts]: Nitpick: would prefer `_` as variable name to signal ignoring.

File: app/services/widgets/settings/alert-box/alert-box-api.ts [hunk 9]
```diff
@@ -406,6 +406,38 @@ interface IAlertBoxRaidSettings {
   raid_variations: IAlertBoxVariation[];
 }
 
+// FACEMASKS
+
```
Comment (blackxored) [inline app/services/widgets/settings/alert-box/alert-box-api.ts]: Nitpick: remove empty line.

File: app/services/widgets/settings/alert-box/index.ts [hunk 10]
```diff
@@ -211,6 +212,10 @@ export class AlertBoxService extends WidgetSettingsService<IAlertBoxData> {
       setting.image_href === '/images/gallery/default.gif'
         ? 'http://uploads.twitchalerts.com/image-defaults/1n9bK4w.gif'
         : setting.image_href;
+    const constrainedDuration =
```
Comment (blackxored) [inline app/services/widgets/settings/alert-box/index.ts]: Can we add comments about this logic, or extract consts? Magic numbers aren't always clear on what they meant or why they were chosen. I'd assume `1000` is ms to sec, but even that might be wrong, the `300` is even more obscure.

Comment (gettinToasty) [reply app/services/widgets/settings/alert-box/index.ts]: 300 is the max duration for alerts, 1000 is ms to sec yea

File: app/components/widgets/AlertBox.vue (hunk 3 above)
Comment (gettinToasty) [reply app/components/widgets/AlertBox.vue]: not sure which brackets you're referring to here

File: app/components/widgets/AlertBox.vue (hunk 4 above)
Comment (gettinToasty) [reply app/components/widgets/AlertBox.vue]: extracting doesnt work here since it's in a for loop

File: app/components/widgets/AlertBox.vue (hunk 5 above)
Comment (gettinToasty) [reply app/components/widgets/AlertBox.vue]: again not sure about which brackets you're referring to... i dont even see any here

File: app/components/widgets/AlertBox.vue.ts (hunk 6 above)
Comment (gettinToasty) [reply app/components/widgets/AlertBox.vue.ts]: it does change

File: app/components/widgets/AlertBox.vue.ts (hunk 7 above)
Comment (gettinToasty) [reply app/components/widgets/AlertBox.vue.ts]: i dont think we report to sentry on this yet, but would be a good thing to add in, i'll make a future PR with that

File: app/services/widgets/settings/alert-box/index.ts (hunk 10 above)
Comment (gettinToasty) [reply app/services/widgets/settings/alert-box/index.ts]: 1000 is ms to sec yea, 300 seconds is the maximum duration for alerts. would really prefer to have things like this handled on the backend but this is the api we inherited

**Pull request merged**